index.js

```diff
@@ -19,20 +20,62 @@ const promisedReadPackage = function (pkgPath) {
   })
 }
 
+const resolveGlobbedPath = function (entries, cwd) {
+  const paths = []
+
+  if (typeof entries === 'string') entries = [entries]
+
+  debug('globby resolving', entries)
+
+  globby.sync(entries, {
+    cwd,
+    absolute: true,
+    expandDirectories: false
+  }).forEach(entry => {
+    // Globby yields unix-style paths.
+    const normalized = path.resolve(entry)
+
+    if (!paths.includes(normalized)) {
+      paths.push(normalized)
+    }
+  })
+
+  debug('globby resolved', paths)
+
+  return paths
+}
+
 module.exports = function (opts, cb) {
   let pkgPath = opts.path
+  let entries
+
   const result = promisedReadPackage(pkgPath)
     .catch(err => {
-      if (err && err.code === 'EISDIR') {
+      if (!err) {
+        return Promise.reject(new Error('Failed to read package.json, but received no error'))
+      } else if (pkgPath.endsWith('/package.json') || pkgPath === 'package.json') {
+        return Promise.reject(new Error('Failed to read package.json: ' + err.message))
+      } else if (err.code === 'EISDIR') {
         pkgPath = path.join(pkgPath, 'package.json')
         return promisedReadPackage(pkgPath)
       }
-      return Promise.reject(err)
+
+      // We've likely been given entries rather than a package.json or module path, try resolving that instead
+      entries = resolveGlobbedPath(pkgPath)
+
+      if (!entries[0]) {
+        return Promise.reject(new Error('Failed to find package.json, could not find any matching files'))
+      }
+
+      opts.noDefaultEntries = true
+      pkgPath = pkgUp.sync({ cwd: path.dirname(entries[0]) })
```

Comment:
Let's avoid synchronous code altogether, `async`/`await` should make this easier too.